reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,399 @@
+2. Improve the expression of NodeAffinity to support affinity\anti-affinity and labels expressions.
+3. The actual principle of NodeAffinitySchedulingStratgy is. According to the label matching expression in the strategy, traverse and search for LabelsResource of all nodes. Pick out nodes that meet the expression requirements.
+
+
+![NodeAffinity Concept](https://user-images.githubusercontent.com/11072802/217133358-19ddc916-bf15-4f69-96e1-9cbd157b0e1d.png)
+
+Scheduling Policy | Label Owner | Label select operator
+-- | -- | --
+NodeAffinity | Node | in, not_in, exists, does_not_exist
+
+### API Design
+
+**The apis of actors/nodes add labels**  
+This interface is already very simple, so we will not set up multiple solutions for everyone to discuss.
+```python
+#  Node add static labels. 
+ray start ... --labels={"location": "dc_1"}
+
+# Dynamically updating node labels is not implemented now, and will be considered in the second stage.
+```
+
+**The apis of the actor-affinity/node-affinity scheduling.**  
+
+**Option 1: Simplify through syntactic sugar**
+
+
+```python
+# Scheduled to a node with a specific IP.
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_in(key="node_ip", values=["xxx.xxx.xx.xx"], is_soft=false))
+    ).remote()
+
+# Try to schedule to the node with A100/P100 graphics card. If not, schedule to other nodes.
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_in("gpu_type", ["A100", "P100"], is_soft=true))
+    ).remote()
+
+# Do not schedule to the two nodes whose node id is "xxxxxxx"\"aaaaaaaa".
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_not_in("node_id", ["xxxxxxx", "aaaaaaaa"], is_soft=false))
+    ).remote()
+
+# Schedule to the node with the key label exist "gpu_type".
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity(label_exist("gpu_type"))
+    ).remote()
+
+# Don't schedule to the node with the key label exist "gpu_type".
+object_ref = Task.options(
+        scheduling_strategy=node_affinity(label_does_not_exist("gpu_type", is_soft=false))
+    ).remote()
+
+# Multiple label expressions can be filled in at the same time, and the relationship between the expressions is "and". The dispatch must satisfy each expression.
+# The actual meaning of this expression is that it must be scheduled to a node with a GPU, and as much as possible to a node with a GPU of the A100 type.
+actor_1 = Actor.options(
+        scheduling_strategy=node_affinity([
+            label_in("gpu_type", ["A100"], true),
+            label_exists("gpu_type", false)
+        ])
+    ).remote()
+
+# Note: This api is old. The old api will still be kept for compatibility with old users.
+actor_1 = Actor.options(
+    scheduling_strategy=NodeAffinitySchedulingStrategy(
+        node_id=ray.get_runtime_context().node_id,
+        soft=False,
+    )
+).remote()
+
+def node_affinity(...):
```

Comment:
Instead of introducing a new `node_affinity` wrapper, we can just swap out the `NodeAffinitySchedulingStrategy` implementation to use labels instead.

Reply:
Same as above explanation. Label expressions are planned to be used in more places in the future. For example, ActorAffinity or get a batch of nodes through label. So it's better to wrap it with node_affinity/NodeAffinitySchedulingStrategy here. Then the label indicates that it can be reused in multiple places.
```
# get a batch of nodes through label.
Nodes = GetNodeInfoByLabel([label_in("zone", ["usa", "china"])
```

Reply:
i think that's option2 right?

Reply:
Not exactly. Option 2 doesn't have `label_in` etc?

Reply:
@jjyao  Do you mean "scheduling_strategy = NodeAffinitySchedulingStrategy([label_in("zone", ["usa", "china"]])?  

NodeAffinitySchedulingStrategy is replaced by node_affinity mainly for more concise and language style more in line with python.

Reply:
@larrylian we can stick with NodeAffinitySchedulingStrategy for the first version, and improve that in a separate PR.

Reply:
@scv119  Do you mean "scheduling_strategy = NodeAffinitySchedulingStrategy([label_in("zone", ["usa", "china"]])" ?    
It's ok to do that。